sdk/eventhub/Azure.Messaging.EventHubs/stress/src/Infrastructure/EventTracking.cs

```diff
@@ -0,0 +1,246 @@
+        {
+            eventData.Properties.Add(PartitionPropertyName, partition);
+        }
+    }
+
+    /// <summary>
+    ///   Processes the <see cref="EventData"/> instance held in the <see cref="ProcessEventArgs"/> in order to determine
+    ///   if the event has already been seen, if the event was received out of order, or if the body is invalid.
+    /// </summary>
+    ///
+    /// <param name="args">The <see cref="ProcessEventArgs"/> received from the processor client to be used for processing.</param>
+    /// <param name="metrics">The <see cref="Metrics"/> instance used to send information about the processed event to Application Insights.</param>
+    ///
+    public async void ProcessEventAsync(ProcessEventArgs args, Metrics metrics)
+    {
+        var indexNumber = _checkEvent(args.Data, args.Partition.PartitionId, metrics);
+
+        if (indexNumber % 100 == 0)
+        {
+            await args.UpdateCheckpointAsync(args.CancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    /// <summary>
+    ///   Processes the <see cref="EventData"/> instance held in the <see cref="PartitionEvent"/> in order to determine
+    ///   if the event has already been seen, if the event was received out of order, or if the body is invalid.
+    /// </summary>
+    ///
+    /// <param name="partitionEvent">The <see cref="PartitionEvent"/> received from the consumer client to be used for processing.</param>
+    /// <param name="metrics">The <see cref="Metrics"/> instance used to send information about the processed event to Application Insights.</param>
+    ///
+    public void ConsumeEvent(PartitionEvent partitionEvent, Metrics metrics) => _checkEvent(partitionEvent.Data, partitionEvent.Partition.PartitionId, metrics);
+
+    /// <summary>
+    ///   Gets the assigned partitions for the <see cref="PartitionPublisher"/>. This allows each partition publisher instance to publish
+    ///   to distinct and separate partitions. This is necessary so that missing and duplicate events can be detected by the <see cref="Role.Consumer"/>
+    ///   role without needing access to the <see cref="PartitionPublisher"/> instance.
+    /// </summary>
+    ///
+    /// <param name="partitionCount"></param>
+    /// <param name="roleIndex"></param>
+    /// <param name="partitionIds"></param>
+    ///
+    /// <returns>The array of partitions that a given <see cref="PartitionPublisher"/> should send to.</returns>
+    ///
+    public static List<string> GetAssignedPartitions(int partitionCount, int roleIndex, string[] partitionIds, Role[] roles)
+    {
+        var roleList = new List<Role>(roles);
+
+        var numPublishers = roleList.Where(role => (role == Role.Publisher || role == Role.BufferedPublisher || role == Role.PartitionPublisher)).Count();
+        var thisPublisherIndex = (roleList.GetRange(0,roleIndex)).Where(role => (role == Role.Publisher || role == Role.BufferedPublisher || role == Role.PartitionPublisher)).Count();
```

Comment:
```suggestion
        var thisPublisherIndex = (roleList.GetRange(0,roleIndex)).Count(role => (role == Role.Publisher || role == Role.BufferedPublisher || role == Role.PartitionPublisher));
```

Reply:
Wouldn't this produce a 1-based index?

Reply:
I could be wrong but I don't think so, GetRange (int index, int count) should look at everything up to but not including the current publisher when you do (0, roleIndex).

Like if you had [x,x,publisher,publisher] with roleIndex 2 that would do GetRange(index: 0, count:2) so that would look at x and x and return 0.